app/services/tracking_service.rb

```
@@ -1,45 +1,78 @@
 class TrackingService
-  def self.track_event(event_text, properties = {})
-    return false unless key.present?
+  TrackingServiceError = Class.new(StandardError)
 
-    client.track_event(event_text, properties: properties)
+  API_ENDPOINT = 'https://www.google-analytics.com/collect'.freeze
+  DEBUG_API_ENDPOINT = 'https://www.google-analytics.com/debug/collect'.freeze
+
+  attr_reader :response, :ga_tracking_id
+
+  def initialize(ga_tracking_id: Rails.configuration.google_analytics_tracking_id, debug: false)
+    @ga_tracking_id = ga_tracking_id
+    @debug = debug
   end
 
-  def self.key
-    @key ||= Rails.configuration.app_insights_instrumentation_key
+  def track_event(key:, label:, value:)
+    return unless ga_tracking_id && [key, label, value].all?
```

I thought you said that label was optional? This is enforcing that all values are passed. In which case why not just make them mandatory parameters and simplify this line?

---

I decided to make the label mandatory within our context.  We won't be sending unlabelled events as it would increase the difficulty for reporting. The label is not mandatory for the GA API, but it is for us.

---

Fair enough, so why not change the first couple of lines to make this super explicit (and simpler implementation):

```ruby
def track_event(:key, :label, :value)
  return unless ga_tracking_id.present?
```

This will then throw an ArgumentError if called without passing all three named arguments - which is preferable to checking within the method that all three were passed.

---

Does it?

```ruby
SyntaxError:
  get-help-to-retrain/app/services/tracking_service.rb:14: syntax error, unexpected tSYMBEG, expecting ')'
    def track_event(:key, :label, :value)
```

---

I had the syntax wrong, but the arguments are still required - hence checking that they are present is redundant. This should work:

```ruby
def track_event(key:, label:, value:)
  return unless ga_tracking_id.present?
```

---

But if you pass

```ruby
track_event(key: nil, label: 1, value: 2)
```
I don't think it will.

---

Hmm this is interesting, I'm not sure failing silently is going to help us here, we need to know something is wrong and if we are sending empty values

---

I will keep it as is now - will think about it when I do the batching.